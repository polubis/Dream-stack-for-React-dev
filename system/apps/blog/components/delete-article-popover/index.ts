export * from './delete-article-popover';
