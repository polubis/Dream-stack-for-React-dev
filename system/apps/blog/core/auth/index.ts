export * from './use-auth';
export * from './guards';
