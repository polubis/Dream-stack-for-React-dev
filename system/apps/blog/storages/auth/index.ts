export * from './auth.storage';
export * from './defs';
