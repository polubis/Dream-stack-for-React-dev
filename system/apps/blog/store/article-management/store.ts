import { create } from 'zustand';
import type * as ArticleManagement from './defs';

const useArticleManagementStore = create<ArticleManagement.State>(() => ({
  is: 'idle',
}));

export { useArticleManagementStore };
