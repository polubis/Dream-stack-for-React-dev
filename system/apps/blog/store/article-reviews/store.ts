import { create } from 'zustand';
import type { ArticleReviewsStore } from './defs';

const useArticleReviewsStore = create<ArticleReviewsStore.State>(() => ({
  is: 'idle',
}));

export { useArticleReviewsStore };
