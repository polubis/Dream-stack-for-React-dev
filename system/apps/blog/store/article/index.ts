export * from './store';
export * from './defs';
export * from './actions';
export * from './selectors';
