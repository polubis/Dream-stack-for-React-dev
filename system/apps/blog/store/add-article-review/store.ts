import { create } from 'zustand';
import type { AddArticleReviewStore } from './defs';
import { addArticleReviewForm } from './core';

const useAddArticleReviewStore = create<AddArticleReviewStore.State>(() => ({
  is: 'idle',
  form: addArticleReviewForm.init({ content: '' }),
}));

export { useAddArticleReviewStore, addArticleReviewForm };
