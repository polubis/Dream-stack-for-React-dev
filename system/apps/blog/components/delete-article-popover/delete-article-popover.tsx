import {
  Box,
  Button,
  CloseIcon,
  Field,
  Font,
  Input,
  Popover,
  TrashIcon,
} from '@system/figa-ui';
import { article_store_selectors } from '../../store/article';
import { useToggle } from '@system/figa-hooks';
import { type FormEventHandler, useState, useEffect } from 'react';
import {
  delete_article_store_actions,
  delete_article_store_selectors,
} from '../../store/delete-article';
import { useRouter } from 'next/router';
import { useLang } from '../../dk';
import { auth_store_selectors } from '../../store/auth';

const Content = () => {
  const { close } = Popover.use();
  const { title, id } = article_store_selectors.useArticle();
  const confirm = useToggle();
  const [currentTitle, setCurrentTitle] = useState('');
  const { is } = delete_article_store_selectors.useState();

  const router = useRouter();
  const lang = useLang();

  const handleSubmit: FormEventHandler<HTMLFormElement> = async (e) => {
    e.preventDefault();

    if (confirm.closed) {
      confirm.open();
      return;
    }

    try {
      await delete_article_store_actions.delete(id);
      router.push(`/${lang}/your-articles`);
    } catch {
      confirm.close();
    }
  };

  const busy = is === 'busy';

  return (
    <Popover.Content
      padding={[250, 250, 250, 250]}
      variant="outlined"
      minWidth="280px"
      maxWidth="440px"
      spacing={[100, 100, 250]}
    >
      <Box orientation="row" between spacing={[250]}>
        <Font variant="h6">Are you sure you want to delete the article?</Font>
        <Button
          size={1}
          shape="rounded"
          variant="outlined"
          motive="tertiary"
          disabled={busy}
          onClick={close}
        >
          <CloseIcon />
        </Button>
      </Box>
      <Font variant="b1">
        This action is{' '}
        <Font element="strong" variant="b1" motive="primary" bold>
          irreversible
        </Font>
        . Ensure you are fully aware of the consequences before proceeding.
      </Font>
      <Font variant="b1">
        Type{' '}
        <Font element="strong" motive="primary" variant="b1">
          &quot;{title}&quot;
        </Font>{' '}
        title to confirm.
      </Font>
      <form onSubmit={handleSubmit}>
        <Box spacing={[300]}>
          <Field
            label="Article title*"
            error={
              is === 'fail' ? 'Something went wrong, try again.' : undefined
            }
          >
            <Input
              placeholder="Type article title to confirm..."
              onChange={(e) => setCurrentTitle(e.target.value)}
            />
          </Field>
          <Box orientation="row" spacing={[150]}>
            <Button
              loading={busy || is === 'ok'}
              type="submit"
              size={2}
              disabled={currentTitle !== title}
            >
              {confirm.opened ? 'Sure?' : 'Delete'}
            </Button>
            <Button
              type="button"
              variant="outlined"
              motive="tertiary"
              size={2}
              disabled={busy}
              onClick={close}
            >
              Cancel
            </Button>
          </Box>
        </Box>
      </form>
    </Popover.Content>
  );
};

const Trigger = () => {
  const { toggle } = Popover.use();

  return (
    <Popover.Trigger>
      <Button
        motive="tertiary"
        title="Delete article"
        shape="rounded"
        size={2}
        onClick={toggle}
      >
        <TrashIcon />
      </Button>
    </Popover.Trigger>
  );
};

const DeleteArticlePopover = () => {
  useEffect(() => {
    return () => {
      delete_article_store_actions.reset();
    };
  }, []);

  return (
    <Popover>
      <Trigger />
      <Content />
    </Popover>
  );
};

const ProtectedDeleteArticlePopover = () => {
  const { authorName } = article_store_selectors.useArticle();
  const isAuthor = auth_store_selectors.useIsAuthor(authorName);
  const isAdmin = auth_store_selectors.useIsAdmin();

  return isAdmin || isAuthor ? <DeleteArticlePopover /> : null;
};

export { ProtectedDeleteArticlePopover as DeleteArticlePopover };
