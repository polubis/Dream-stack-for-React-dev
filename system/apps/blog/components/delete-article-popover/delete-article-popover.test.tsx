import { act, render, screen } from '@testing-library/react';
import { DeleteArticlePopover } from './delete-article-popover';
import { storeFixture } from '../../store/test-utils';
import { useArticleStore } from '../../store/article';
import { useAuthStore } from '../../store/auth';
import {
  mockGetArticleResponse,
  mockSignedInUser,
} from '@system/blog-api-mocks';
import { useRouter } from 'next/router';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

describe('Delete article feature works when: ', () => {
  it('only admin or author of article is able to delete', () => {
    (useRouter as jest.Mock).mockReturnValue({
      pathname: '/en/something/',
    });
    const user = mockSignedInUser();
    const article = mockGetArticleResponse().data;
    const articleStore = storeFixture(useArticleStore, {
      is: 'ok',
      article,
    });
    const authStore = storeFixture(useAuthStore, {
      key: 'authorized',
      user,
      check: jest.fn(),
    });

    const { asFragment, unmount } = render(<DeleteArticlePopover />);

    screen.getByTitle(/Delete article/);

    act(() => {
      useAuthStore.setState({
        key: 'authorized',
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        user: { ...user, roles: ['User'] as any },
        check: jest.fn(),
      });
      useArticleStore.setState({
        is: 'ok',
        article: {
          ...article,
          authorName: 'Some random username',
        },
      });
    });

    expect(screen.queryByTitle(/Delete article/)).toBeFalsy();

    act(() => {
      useAuthStore.setState({
        key: 'authorized',
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        user: { ...user, roles: ['User'] as any },
        check: jest.fn(),
      });
      useArticleStore.setState({
        is: 'ok',
        article: {
          ...article,
          authorName: user.username,
        },
      });
    });

    screen.getByTitle(/Delete article/);
    expect(asFragment()).toMatchSnapshot();

    unmount();
    articleStore.restore();
    authStore.restore();
  });
});
