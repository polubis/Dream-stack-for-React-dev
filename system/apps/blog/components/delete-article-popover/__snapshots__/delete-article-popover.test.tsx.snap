// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete article feature works when:  only admin or author of article is able to delete 1`] = `
<DocumentFragment>
  <div
    class="popover-trigger"
    id="popover-trigger-:r1:"
  >
    <button
      class="button size-2 rounded filled primary"
      shape="rounded"
      size="2"
      title="Delete article"
    >
      <svg
        class="icon trash-icon"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 21C6.45 21 5.979 20.804 5.587 20.412C5.195 20.02 4.99933 19.5493 5 19V6H4V4H9V3H15V4H20V6H19V19C19 19.55 18.804 20.021 18.412 20.413C18.02 20.805 17.5493 21.0007 17 21H7ZM9 17H11V8H9V17ZM13 17H15V8H13V17Z"
          fill="black"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
